src/processor/operator/persistent/csv_reader.cpp

```diff
@@ -0,0 +1,529 @@
+    case LogicalTypeID::FIXED_LIST: {
+        auto fixedListVal = storage::TableCopyUtils::getArrowFixedListVal(
+            strVal, 1, strVal.length() - 2, type, csvReaderConfig);
+        vector->copyFromValue(rowToAdd, *fixedListVal);
+    } break;
+    case LogicalTypeID::STRUCT: {
+        auto structString = strVal.substr(1, strVal.length() - 2);
+        auto structFieldIdxAndValuePairs = storage::TableCopyUtils::parseStructFieldNameAndValues(
+            type, structString, csvReaderConfig);
+        for (auto& fieldIdxAndValue : structFieldIdxAndValuePairs) {
+            copyStringToVector(
+                StructVector::getFieldVector(vector, fieldIdxAndValue.fieldIdx).get(),
+                fieldIdxAndValue.fieldValue);
+        }
+    } break;
+    case LogicalTypeID::UNION: {
+        union_field_idx_t selectedFieldIdx = INVALID_STRUCT_FIELD_IDX;
+        for (auto i = 0u; i < UnionType::getNumFields(&type); i++) {
+            auto internalFieldIdx = UnionType::getInternalFieldIdx(i);
+            if (storage::TableCopyUtils::tryCast(
+                    *UnionType::getFieldType(&type, i), strVal.c_str(), strVal.length())) {
+                StructVector::getFieldVector(vector, internalFieldIdx)
+                    ->setNull(rowToAdd, false /* isNull */);
+                copyStringToVector(
+                    StructVector::getFieldVector(vector, internalFieldIdx).get(), strVal);
+                selectedFieldIdx = i;
+                break;
+            } else {
+                StructVector::getFieldVector(vector, internalFieldIdx)
+                    ->setNull(rowToAdd, true /* isNull */);
+            }
+        }
+        if (selectedFieldIdx == INVALID_STRUCT_FIELD_IDX) {
+            throw ParserException{
+                StringUtils::string_format("No parsing rule matches value: {}.", strVal)};
+        }
+        StructVector::getFieldVector(vector, UnionType::TAG_FIELD_IDX)
+            ->setValue(rowToAdd, selectedFieldIdx);
+        StructVector::getFieldVector(vector, UnionType::TAG_FIELD_IDX)
+            ->setNull(rowToAdd, false /* isNull */);
+    } break;
+    default: {
+        throw NotImplementedException("BaseCSVReader::AddValue");
+    }
+    }
+}
+
+BufferedCSVReader::BufferedCSVReader(const std::string& filePath,
+    common::CSVReaderConfig csvReaderConfig, catalog::TableSchema* tableSchema)
+    : BaseCSVReader{filePath, csvReaderConfig, tableSchema}, bufferSize{0}, position{0}, start{0} {
+    Initialize();
+}
+
+BufferedCSVReader::~BufferedCSVReader() {
+    if (fd != -1) {
+        close(fd);
+    }
+}
+
+void BufferedCSVReader::Initialize() {
+    numPropertiesToCopy = 0;
+    for (auto& property : tableSchema->properties) {
+        if (property->getDataType()->getLogicalTypeID() != common::LogicalTypeID::SERIAL) {
+            numPropertiesToCopy++;
+        }
+    }
+    // TODO(Ziyi): should we wrap this fd using kuzu file handler?
+    fd = open(filePath.c_str(), O_RDONLY);
```

Comment:
We should reuse `FileUtils::openFile` if possible. The returned FileInfo will auto close file.